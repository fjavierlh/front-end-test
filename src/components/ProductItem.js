import React from 'react';
import { Link } from 'react-router-dom';
import styled from 'styled-components';
import { Image } from './Image';

const ProductItem = ({ id, imgUrl, brand, model, price }) => {
  return (
    <ProductItemStyled>
      <Link
        to={`/product/${model}`.toLowerCase().replaceAll(' ', '-')}
        className="product-link"
        key={id}
        state={{ id, imgUrl, brand, model }}
      >
        <Image
          className="img-item"
          imgUrl={imgUrl}
          altText={`${brand}-${model}`}
          lazyLoading={true}
        />
        <div className="product-info">
          <p>
            <b>{brand}</b> {model}
          </p>
          <span className="price-text">
            {price ? <b>{price}â‚¬</b> : <small>Price to be consulted</small>}
          </span>
        </div>
      </Link>
    </ProductItemStyled>
  );
};

const ProductItemStyled = styled.article`
  flex: 1 0 21%;
  min-height: 180px;
  & > .product-link {
    flex: 1 0 21%;
    flex-basis: 100%;
    text-decoration: none;
    color: #333;

    & > .product-info {
      display: flex;
      flex-direction: column;
      justify-content: flex-end;
    }
  }
  text-align: center;
  background-color: #fff;
  padding: 2rem 0 2rem 0;

  & > .img-item {
    display: block;
    max-height: 180px;
    min-height: 180px;

    margin: 0 2em;
  }

  & > .product-info {
    display: flex;
    flex-direction: column;
    justify-content: flex-end;
  }
`;

export default ProductItem;
