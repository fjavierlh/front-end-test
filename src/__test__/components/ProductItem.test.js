import { render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import ProductItem from '../../components/ProductItem';
import { allProducts } from './../../../fixtures/allProducts.fixture';

describe('ProductItem', () => {
  test('should render ProductItem with all passed props', () => {
    const { id, imgUrl, brand, model, price } = allProducts[0];
    render(
      <BrowserRouter>
        <ProductItem
          id={id}
          imgUrl={imgUrl}
          brand={brand}
          model={model}
          price={price}
        />
      </BrowserRouter>
    );
    expect(screen.getByText(brand)).toBeInTheDocument();
    expect(screen.getByText(model)).toBeInTheDocument();
    expect(screen.getByText(`${price}â‚¬`)).toBeInTheDocument();
    expect(screen.getByRole('img')).toBeInTheDocument();
  });

  test('should render ProductItem with default text if price is not passed', () => {
    const { id, imgUrl, brand, model } = allProducts[0];
    const expectedText = 'Price to be consulted';
    render(
      <BrowserRouter>
        <ProductItem id={id} imgUrl={imgUrl} brand={brand} model={model} />
      </BrowserRouter>
    );
    expect(screen.getByText(expectedText)).toBeInTheDocument();
  });

  test('should match snapshot ProductItem with default text if price is not passed', () => {
    const { id, imgUrl, brand, model } = allProducts[0];
    const { container } = render(
      <BrowserRouter>
        <ProductItem id={id} imgUrl={imgUrl} brand={brand} model={model} />
      </BrowserRouter>
    );
    expect(container).toMatchSnapshot();
  });

  test('should match snapshot ProductItem with all passed props', () => {
    console.log(`allProducts`, allProducts[0]);
    const { id, imgUrl, brand, model, price } = allProducts[0];
    const { container } = render(
      <BrowserRouter>
        <ProductItem
          id={id}
          imgUrl={imgUrl}
          brand={brand}
          model={model}
          price={price}
        />
      </BrowserRouter>
    );
    expect(container).toMatchSnapshot();
  });
});
