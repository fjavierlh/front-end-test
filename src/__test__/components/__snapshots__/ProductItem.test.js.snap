// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductItem should match snapshot ProductItem with all passed props 1`] = `
<div>
  <article
    class="sc-bdvvtL kQKxTo"
  >
    <a
      class="product-link"
      href="/product/iconia-talk-s"
    >
      <img
        alt="Acer-Iconia Talk S"
        loading="lazy"
        src="https://front-test-api.herokuapp.com/images/ZmGrkLRPXOTpxsU4jjAcv.jpg"
      />
      <div
        class="product-info"
      >
        <p>
          <b>
            Acer
          </b>
           
          Iconia Talk S
        </p>
        <span
          class="price-text"
        >
          <b>
            170
            â‚¬
          </b>
        </span>
      </div>
    </a>
  </article>
</div>
`;

exports[`ProductItem should match snapshot ProductItem with default text if price is not passed 1`] = `
<div>
  <article
    class="sc-bdvvtL kQKxTo"
  >
    <a
      class="product-link"
      href="/product/iconia-talk-s"
    >
      <img
        alt="Acer-Iconia Talk S"
        loading="lazy"
        src="https://front-test-api.herokuapp.com/images/ZmGrkLRPXOTpxsU4jjAcv.jpg"
      />
      <div
        class="product-info"
      >
        <p>
          <b>
            Acer
          </b>
           
          Iconia Talk S
        </p>
        <span
          class="price-text"
        >
          <small>
            Price to be consulted
          </small>
        </span>
      </div>
    </a>
  </article>
</div>
`;
